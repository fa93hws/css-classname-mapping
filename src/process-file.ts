import { existsSync as _existsSync, readFileSync as _readFileSync } from 'fs';
import { Ok, Err, Result } from 'ts-results';

type FsStub = {
  readFileSync?: typeof _readFileSync;
  existsSync?: typeof _existsSync;
};

export function findSourceMapPath(content: string): Result<string, null> {
  const regex = /^\/\*# sourceMappingURL=(.+)\*\//m;
  const match = regex.exec(content);
  if (match == null || match.length !== 2) {
    return Err(null);
  }
  return Ok(match[1]);
}

export function processFile(
  filePath: string,
  { readFileSync = _readFileSync, existsSync = _existsSync }: FsStub = {},
): Result<string, string> {
  const content = readFileSync(filePath, { encoding: 'utf-8' });
  const sourcemapPathResult = findSourceMapPath(content);
  if (sourcemapPathResult.err) {
    return Err(`Fail to find sourcemap in ${filePath}`);
  }
  const sourcemapPath = sourcemapPathResult.val;
  if (!existsSync(sourcemapPath)) {
    return Err(`sourcemap@${filePath} does not exist`);
  }
  const sourcemapContent = readFileSync(sourcemapPath, { encoding: 'utf-8' });
  return Ok(sourcemapContent);
}
