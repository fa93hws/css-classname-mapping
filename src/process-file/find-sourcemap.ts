import { Ok, Err, Result } from 'ts-results';

export function findSourceMap(content: string): Result<string, null> {
  const regex = /^\/\*\# sourceMappingURL=(.+)\*\//m;
  const match = regex.exec(content);
  if (match == null || match.length !== 2) {
    return Err(null);
  }
  return Ok(match[1]);
}