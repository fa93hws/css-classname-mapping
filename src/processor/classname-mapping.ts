const SEPARATOR = '(ﾟДﾟ)';
const VERSION = 'v1';

type MinifiedClassname = string;
type Value = {
  classname: string;
  source: string;
};
type ValueWithKey = Value & {
  // Serialized key with Sseparator so that we can achieve deduplication via new Set()
  uniqueKey: string;
};
export type OutputObject = {
  version: string;
  mapping: Record<MinifiedClassname, Value[]>;
};

/**
 * In order to deduplication the value based on sourceFile and originalClassname
 * Serialization and deserialization to/from string are introduced
 */
export class ClassnameMapping {
  private mapWithSerializedValue = new Map<MinifiedClassname, ValueWithKey[]>();

  private static serialize({
    classname,
    sourceFile,
  }: {
    classname: string;
    sourceFile: string;
  }): string {
    return classname + SEPARATOR + sourceFile;
  }

  private static deduplicate(valueWithKeys: ValueWithKey[]): Value[] {
    const tempMap = new Map<string, Value>();
    valueWithKeys.forEach(({ uniqueKey, ...value }) => {
      tempMap.set(uniqueKey, value);
    });
    const values: Value[] = [];
    tempMap.forEach((value) => values.push(value));
    return values;
  }

  set({
    minified,
    original,
    sourceFile,
  }: {
    minified: string;
    original: string;
    sourceFile: string;
  }): void {
    const serializedValue = ClassnameMapping.serialize({
      classname: original,
      sourceFile,
    });
    const maybeSerializedValue = this.mapWithSerializedValue.get(minified);
    const valueWithKey: ValueWithKey = {
      classname: original,
      source: sourceFile,
      uniqueKey: serializedValue,
    };
    if (maybeSerializedValue == null) {
      this.mapWithSerializedValue.set(minified, [valueWithKey]);
    } else {
      maybeSerializedValue.push(valueWithKey);
    }
  }

  toOutputObject(): OutputObject {
    const mapping: OutputObject['mapping'] = {};
    this.mapWithSerializedValue.forEach((valueWithKeys, minifiedClassname) => {
      mapping[minifiedClassname] = ClassnameMapping.deduplicate(valueWithKeys);
    });
    return {
      version: VERSION,
      mapping,
    };
  }
}
