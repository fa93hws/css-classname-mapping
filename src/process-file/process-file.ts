import { readFileSync as _readFileSync } from 'fs';
import { findSourceMap } from './find-sourcemap';

export function processFile(
  filePath: string,
  {
    readFileSync = _readFileSync,
  }: {
    readFileSync?: typeof _readFileSync;
  } = {},
) {
  const content = readFileSync(filePath, { encoding: 'utf-8' });
  const sourcemapPath = findSourceMap(content);
}
