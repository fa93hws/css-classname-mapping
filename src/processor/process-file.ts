import { RawSourceMap, SourceMapConsumer } from 'source-map';

import { pairSelectors } from './selectors-pairing';
import { ClassnameMapping, OutputObject } from './classname-mapping';
import { CssParser } from '../utils/css-parser';
import { Location } from './location';
import { findSourceMap } from './find-sourcemap';
import {
  generateLocationToSelector,
  LocationToSelector,
} from './selector-location';

async function getOriginalSelectors(
  serializedLoc: string,
  sourcemap: RawSourceMap,
  // source -> Map<location serielized key, classname>
  sourceLocToSelector: Map<string, LocationToSelector>,
): Promise<{
  selector: string;
  file: string;
}> {
  return new Promise((resolve, reject) => {
    const location = Location.fromSerializedKey(serializedLoc);
    SourceMapConsumer.with(sourcemap, null, (consumer) => {
      const {
        source: sourcePath,
        line: sourceLine,
        column: sourceColumn,
      } = consumer.originalPositionFor(location);
      if (sourcePath == null || sourceLine == null || sourceColumn == null) {
        return reject(
          new Error(`fail to find the original location @${serializedLoc}`),
        );
      }
      const locationToSelector = sourceLocToSelector.get(sourcePath);
      if (locationToSelector == null) {
        return reject(
          new Error(
            `fail to get location to selector mapping for ${sourcePath}`,
          ),
        );
      }
      const sourceLocation = new Location({
        line: sourceLine,
        column: sourceColumn,
      });
      const originalSelector = locationToSelector.get(
        sourceLocation.serialize(),
      );
      if (originalSelector == null) {
        return reject(
          new Error(
            `fail to get original selector@${sourceLocation.serialize()} in ${sourceLocation}`,
          ),
        );
      }
      return resolve({
        selector: originalSelector,
        file: sourcePath,
      });
    });
  });
}

function generateLocationToSelectorFromSources(
  sourcemap: RawSourceMap,
  cssParser: CssParser,
): Map<string, LocationToSelector> {
  const output = new Map<string, LocationToSelector>();
  const { sources: paths, sourcesContent: contents } = sourcemap;
  if (contents == null) {
    throw new Error('content can not be empty');
  }
  if (paths.length !== contents.length) {
    throw new Error('lengh of the paths and contents do not match');
  }
  for (let idx = 0; idx < paths.length; idx += 1) {
    const path = paths[idx];
    const rootNode = cssParser.parseFromContent({
      content: contents[idx],
      cacheKey: path,
    });
    const locationToSelector = generateLocationToSelector({
      rootNode,
      cacheKey: path,
    });
    output.set(path, locationToSelector);
  }
  return output;
}

async function generateMapping(
  minifiedLocToSelector: LocationToSelector,
  sourceLocToSelector: Map<string, LocationToSelector>,
  sourcemap: RawSourceMap,
): Promise<OutputObject> {
  const classnameMapping = new ClassnameMapping();
  const promises = Array.from(minifiedLocToSelector).map(
    async ([minifiedLoc, minifiedSelector]) => {
      const originalSelector = await getOriginalSelectors(
        minifiedLoc,
        sourcemap,
        sourceLocToSelector,
      );
      const localMapping = pairSelectors({
        minified: minifiedSelector,
        original: originalSelector.selector,
      });
      localMapping.forEach((originalClassnames, minifiedClassname) => {
        originalClassnames.forEach((originalClassname) => {
          classnameMapping.set({
            minified: minifiedClassname,
            original: originalClassname,
            sourceFile: originalSelector.file,
          });
        });
      });
    },
  );
  await Promise.all(promises);
  return classnameMapping.toOutputObject();
}

export async function processFile(cssFilePath: string): Promise<void> {
  const cssParser = new CssParser();
  const { css, sourcemap } = findSourceMap(cssFilePath).unwrap();
  const minifiedCssRootNode = cssParser.parseFromContent({
    content: css.content,
    cacheKey: css.path,
  });
  const minifiedLocToSelector = generateLocationToSelector({
    rootNode: minifiedCssRootNode,
    cacheKey: css.path,
  });
  const sourceLocToSelector = generateLocationToSelectorFromSources(
    sourcemap.raw,
    cssParser,
  );
  try {
    const output = await generateMapping(
      minifiedLocToSelector,
      sourceLocToSelector,
      sourcemap.raw,
    );
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(`fail to process file ${cssFilePath}`);
    // eslint-disable-next-line no-console
    console.error(e);
    process.exit(1);
  }
}
