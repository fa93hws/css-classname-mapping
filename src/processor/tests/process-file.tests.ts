import path from 'path';
import semver from 'semver';

import { ClassnameMapping } from '../classname-mapping';
import { processFile } from '../process-file';

describe('processFile', () => {
  it('generate version according to major version of this library', async () => {
    const mapping = new ClassnameMapping();
    await processFile(
      path.join(__dirname, 'fixtures', 'single_file.css'),
      mapping,
    );
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const thisLibraryVersion = require('../../../package.json').version;
    expect(mapping.toOutputObject().version).toEqual(
      `v${semver.parse(thisLibraryVersion)?.major}`,
    );
  });

  it('generate mapping for one file', async () => {
    const mapping = new ClassnameMapping();
    await processFile(
      path.join(__dirname, 'fixtures', 'single_file.css'),
      mapping,
    );
    expect(mapping.toOutputObject().mapping).toMatchSnapshot();
  });

  it('generate mapping for multiple files', async () => {
    const mapping = new ClassnameMapping();
    await processFile(
      path.join(__dirname, 'fixtures', 'single_file.css'),
      mapping,
    );
    await processFile(
      path.join(__dirname, 'fixtures', 'multiple_files.css'),
      mapping,
    );
    await processFile(
      path.join(__dirname, 'fixtures', 'main.d4cf772d.css'),
      mapping,
    );
    expect(mapping.toOutputObject().mapping).toMatchSnapshot();
  });

  it('aggregate mappings for multiple files', async () => {
    const mapping1 = new ClassnameMapping();
    await processFile(
      path.join(__dirname, 'fixtures', 'single_file.css'),
      mapping1,
    );
    const mapping2 = new ClassnameMapping();
    await processFile(
      path.join(__dirname, 'fixtures', 'single_file.css'),
      mapping2,
    );
    await processFile(
      path.join(__dirname, 'fixtures', 'single_file.css'),
      mapping2,
    );
    expect(mapping1.toOutputObject()).toEqual(mapping2.toOutputObject());
  });
});
