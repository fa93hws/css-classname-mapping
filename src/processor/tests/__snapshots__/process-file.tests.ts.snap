// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processFile generate mapping for multiple files 1`] = `
Object {
  "BFpQDNJ62UOlaB933I4jA": Array [
    Object {
      "classname": "emptyCartWrapper",
      "source": "webpack:///cart.css",
    },
  ],
  "DmVGGw": Array [
    Object {
      "classname": "dog",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "E9sfBenJb2vrfMALrKIwF": Array [
    Object {
      "classname": "itemImg",
      "source": "webpack:///menu-items.css",
    },
  ],
  "FHa0MQ": Array [
    Object {
      "classname": "foo",
      "source": "webpack:///pages/dummy/fakes/foo.css",
    },
  ],
  "GWn3Lg": Array [
    Object {
      "classname": "boy",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "HHNCVg": Array [
    Object {
      "classname": "fakeComponent",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
    Object {
      "classname": "fakeComponent",
      "source": "webpack:///pages/dummy/fakes/fake_component2.css",
    },
  ],
  "IGlTFQ": Array [
    Object {
      "classname": "gold",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "JYYVmvVrh64RkTqH6K0hD": Array [
    Object {
      "classname": "img",
      "source": "webpack:///cart.css",
    },
  ],
  "P28ZkrQteFEke19CiGj9m": Array [
    Object {
      "classname": "nameContainer",
      "source": "webpack:///cart.css",
    },
  ],
  "R7j1Nfa3LjNycEIPp_iDH": Array [
    Object {
      "classname": "menuListWrapper",
      "source": "webpack:///menu.css",
    },
  ],
  "Rebjuw": Array [
    Object {
      "classname": "bar",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "Tt7OJQ": Array [
    Object {
      "classname": "foo",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "U903it5lvim_DL1fCiq3f": Array [
    Object {
      "classname": "carouselContainer",
      "source": "webpack:///carousel.css",
    },
  ],
  "ZHL81Jgo4v7QBES0IrBP8": Array [
    Object {
      "classname": "settlementButtonWrapper",
      "source": "webpack:///cart.css",
    },
  ],
  "ZlsV-AlqhtczpPaE09QdU": Array [
    Object {
      "classname": "quantityWrapper",
      "source": "webpack:///menu-items.css",
    },
  ],
  "_-di0Q": Array [
    Object {
      "classname": "bar",
      "source": "webpack:///pages/dummy/fakes/bar.css",
    },
  ],
  "_0qzQbw": Array [
    Object {
      "classname": "baz",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "_1A1J4xlpKobkX8JfUa45uQ": Array [
    Object {
      "classname": "paymentContainer",
      "source": "webpack:///cart.css",
    },
  ],
  "_1A7462naFMcSgjKyEsKjS9": Array [
    Object {
      "classname": "buttonWrapper",
      "source": "webpack:///menu.css",
    },
  ],
  "_1A94Db0ZH6ozSo4Y5upENp": Array [
    Object {
      "classname": "cardSectionContainer",
      "source": "webpack:///menu.css",
    },
  ],
  "_1VuT3inJLojeteNNdUvSeh": Array [
    Object {
      "classname": "container",
      "source": "webpack:///cart.css",
    },
  ],
  "_1fPd-7XrFfpuuLiqIhZNih": Array [
    Object {
      "classname": "subtitleWrapper",
      "source": "webpack:///dish-grid.css",
    },
  ],
  "_1jY1h2qego6cmSVUV4c6oW": Array [
    Object {
      "classname": "itemWrapper",
      "source": "webpack:///menu-items.css",
    },
  ],
  "_2-h4JyPN1zty-2ezUETtgX": Array [
    Object {
      "classname": "menuItemsWrapper",
      "source": "webpack:///menu.css",
    },
  ],
  "_20YIbCYwozfR1crcXtf74J": Array [
    Object {
      "classname": "image",
      "source": "webpack:///carousel.css",
    },
  ],
  "_2AzRPdTLc3tYTTqYfOcRn3": Array [
    Object {
      "classname": "cartIconWrapper",
      "source": "webpack:///menu.css",
    },
  ],
  "_2Lo9r3DIpLfXINNKXdnqin": Array [
    Object {
      "classname": "titleWrapper",
      "source": "webpack:///category.css",
    },
  ],
  "_2NW5Gqlp6U0MxMHZF8Cyn0": Array [
    Object {
      "classname": "priceWrapper",
      "source": "webpack:///menu.css",
    },
  ],
  "_2l-73rAeK9HNiIdxNyHGq-": Array [
    Object {
      "classname": "quantityControlWrapper",
      "source": "webpack:///cart.css",
    },
  ],
  "_2ll7Xt6a1l-Ihmzih7_ceO": Array [
    Object {
      "classname": "container",
      "source": "webpack:///menu.css",
    },
  ],
  "_2ne6-q6N6YmOSqGj7y-Re_": Array [
    Object {
      "classname": "imgContainer",
      "source": "webpack:///cart.css",
    },
  ],
  "_2tbeS65Uv9p8YgM1sDkKbA": Array [
    Object {
      "classname": "menuWrapper",
      "source": "webpack:///menu.css",
    },
  ],
  "_35-2JpizhBHmYg-1nBBIk2": Array [
    Object {
      "classname": "quantityControlItem",
      "source": "webpack:///cart.css",
    },
  ],
  "_3N5ioClhweSODWrnZSzebk": Array [
    Object {
      "classname": "mainContainer",
      "source": "webpack:///global.css",
    },
  ],
  "_3PsbIE_FLVJ3gR42_KuxFf": Array [
    Object {
      "classname": "categoryIcon",
      "source": "webpack:///home.css",
    },
  ],
  "_3QSbfmDpc-fDNTOFkZSi-k": Array [
    Object {
      "classname": "controlWrapper",
      "source": "webpack:///dish-grid.css",
    },
  ],
  "_3R04rJ5KZsvDV4l0ilrzFt": Array [
    Object {
      "classname": "carouselWrapper",
      "source": "webpack:///carousel.css",
    },
  ],
  "_3UhAfYE8ikBH_9WJZezPdi": Array [
    Object {
      "classname": "link",
      "source": "webpack:///home.css",
    },
  ],
  "_3cyTRp_QEE8P3iHh4Sk6xG": Array [
    Object {
      "classname": "carouselItem",
      "source": "webpack:///carousel.css",
    },
  ],
  "_3nUa7Vg_7poLNWjvPUZKNf": Array [
    Object {
      "classname": "rootContainer",
      "source": "webpack:///global.css",
    },
  ],
  "_Zwvmw": Array [
    Object {
      "classname": "apple",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "auy1LigLCFwil0QvzoQJf": Array [
    Object {
      "classname": "itemTextWrapper",
      "source": "webpack:///menu-items.css",
    },
  ],
  "ce-FMt6gHZosmh0dV2SCM": Array [
    Object {
      "classname": "categoryItem",
      "source": "webpack:///home.css",
    },
  ],
  "lXfJHQ": Array [
    Object {
      "classname": "ele",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "tHNuqw": Array [
    Object {
      "classname": "cat",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
}
`;

exports[`processFile generate mapping for one file 1`] = `
Object {
  "DmVGGw": Array [
    Object {
      "classname": "dog",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "GWn3Lg": Array [
    Object {
      "classname": "boy",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "HHNCVg": Array [
    Object {
      "classname": "fakeComponent",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "IGlTFQ": Array [
    Object {
      "classname": "gold",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "Rebjuw": Array [
    Object {
      "classname": "bar",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "Tt7OJQ": Array [
    Object {
      "classname": "foo",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "_0qzQbw": Array [
    Object {
      "classname": "baz",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "_Zwvmw": Array [
    Object {
      "classname": "apple",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "lXfJHQ": Array [
    Object {
      "classname": "ele",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
  "tHNuqw": Array [
    Object {
      "classname": "cat",
      "source": "webpack:///pages/dummy/fakes/fake_component.css",
    },
  ],
}
`;
